backends/hip-ref/ceed-hip-ref-operator.c

```diff
@@ -1239,6 +1248,353 @@ static int CeedOperatorLinearAssembleAddPointBlockDiagonal_Hip(CeedOperator op,
+      CeedChk(ierr);
+      if (eval_mode != CEED_EVAL_NONE) {
+        ierr = CeedRealloc(num_B_in_mats_to_load + 1, &eval_mode_in); CeedChk(ierr);
+        eval_mode_in[num_B_in_mats_to_load] = eval_mode;
+        num_B_in_mats_to_load += 1;
+        if (eval_mode == CEED_EVAL_GRAD) {
+          num_emode_in += dim;
+          size_B_in += dim * esize * nqpts;
+        } else {
+          num_emode_in +=1;
+          size_B_in += esize * nqpts;
+        }
+      }
+    }
+  }
+
+  // Determine active output basis; basis_out and rstr_out only used if same as input, TODO
+  ierr = CeedQFunctionGetFields(qf, NULL, NULL, NULL, &qf_fields); CeedChk(ierr);
+  CeedInt num_emode_out = 0, num_B_out_mats_to_load = 0, size_B_out = 0;
+  CeedEvalMode *eval_mode_out = NULL;
+  CeedBasis basis_out = NULL;
+  CeedElemRestriction rstr_out = NULL;
+  for (CeedInt i=0; i<num_output_fields; i++) {
+    CeedVector vec;
+    ierr = CeedOperatorFieldGetVector(output_fields[i], &vec); CeedChk(ierr);
+    if (vec == CEED_VECTOR_ACTIVE) {
+      ierr = CeedOperatorFieldGetBasis(output_fields[i], &basis_out);
+      CeedChk(ierr);
+      ierr = CeedOperatorFieldGetElemRestriction(output_fields[i], &rstr_out);
+      CeedChk(ierr);
+      if (rstr_out && rstr_out != rstr_in)
+        // LCOV_EXCL_START
+        return CeedError(ceed, CEED_ERROR_BACKEND,
+                         "Multi-field non-composite operator assembly not supported");
+      // LCOV_EXCL_STOP
+      CeedEvalMode eval_mode;
+      ierr = CeedQFunctionFieldGetEvalMode(qf_fields[i], &eval_mode);
+      CeedChk(ierr);
+      if (eval_mode != CEED_EVAL_NONE) {
+        ierr = CeedRealloc(num_B_out_mats_to_load + 1, &eval_mode_out); CeedChk(ierr);
+        eval_mode_out[num_B_out_mats_to_load] = eval_mode;
+        num_B_out_mats_to_load += 1;
+        if (eval_mode == CEED_EVAL_GRAD) {
+          num_emode_out += dim;
+          size_B_out += dim * esize * nqpts;
+        } else {
+          num_emode_out +=1;
+          size_B_out += esize * nqpts;
+        }
+      }
+    }
+  }
+
+  if (num_emode_in == 0 || num_emode_out == 0)
+    // LCOV_EXCL_START
+    return CeedError(ceed, CEED_ERROR_UNSUPPORTED,
+                     "Cannot assemble operator without inputs/outputs");
+  // LCOV_EXCL_STOP
+
+  CeedInt nelem, ncomp;
+  ierr = CeedElemRestrictionGetNumElements(rstr_in, &nelem); CeedChk(ierr);
+  ierr = CeedElemRestrictionGetNumComponents(rstr_in, &ncomp); CeedChk(ierr);
+
+  ierr = CeedCalloc(1, &impl->asmb); CeedChkBackend(ierr);
+  CeedOperatorAssemble_Hip *asmb = impl->asmb;
+  asmb->nelem = nelem;
+  asmb->nnodes = esize;
+
+  // Compile kernels
+  int elemsPerBlock = 1;
+  asmb->elemsPerBlock = elemsPerBlock;
+  // TODO: 1D threadblock version for larger elements
+  if (esize * esize * elemsPerBlock > 1024)
+    // LCOV_EXCL_START
+    return CeedError(ceed, CEED_ERROR_UNSUPPORTED,
+                     "GPU assembly not currently available for elements of this size");
+  // LCOV_EXCL_STOP
```

Comment:
maybe something like this?

```suggestion
  if (esize * esize * elemsPerBlock > 1024) {
    // Do some cleanup here
    /* ToDo */

  ierr = CeedSetBackendFunction(ceed, "Operator", op,
                                "LinearAssemble", NULL); CeedChkBackend(ierr);
  // Set some flag to re-try interface level impl and return
  }

// ...

// Back in CeedOperatorLinearAssemble_Hip
// if falling back
ierr = CeedOperatorLinearAssemble(op, values); CeedChkBackend(ierr);
return CEED_ERROR_SUCCESS;
```

Reply:
If we do this, we probably want to include a `CeedDebug256` statement indicating we're falling back.